dotcli-sync workflow fails on fresh pushes due to missing commit SHA (git diff failure)

### Problem Statement

The default GitHub Actions workflow (main.yml) provided in the dotCMS core repository fails during the sync-with-dotcms job when pushing to a newly created repo or after a force push. The failure stems from a git diff command that references a non-existent commit SHA (github.event.before), causing the workflow to exit with code 128.

### Steps to Reproduce

1. Clone the dotCMS core repo and set up GitHub Actions as instructed.


2. Push changes to a new or freshly cloned repo.

Observe the workflow fail on the Get changes step with a message like:

 fatal: bad object [SHA]
Error: Process completed with exit code 128.


### Acceptance Criteria

Root Cause:
 GitHub’s checkout@v4 uses a shallow clone by default (fetch-depth: 1), which doesn't include the previous commit referenced by ${{ github.event.before }}. This results in the SHA not being found in the local Git context.

Fix / Workaround:
 Two changes were required in our main.yml to resolve the issue:

1st change:

Update Checkout Step in sync-with-dotcms:

 - name: Checkout
  uses: actions/checkout@v4
  with:
    fetch-depth: 0

This ensures the full commit history is available.

2nd change:

Add fallback for the git diff step:

 - name: Get changes
  run: |
    if git cat-file -e "${{ github.event.before }}^{commit}"; then
      git diff --name-only "${{ github.event.before }}" "${{ github.event.after }}" > changed_files.txt
    else
      echo "Previous commit not found. Using fallback diff from HEAD^."
      git diff --name-only HEAD^ HEAD > changed_files.txt
    fi

It would be helpful to include this logic in the official template to support both fresh repositories and standard development flows without requiring manual modification of the workflow.

### dotCMS Version

Environment Info:
dotCMS version: 25.07.24-1


dotCLI version: 25.07.24-1


GitHub Runner: ubuntu-24.04


### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

--- a/tools/dotcms-cli/action/.github/workflows/main.yml
+++ b/tools/dotcms-cli/action/.github/workflows/main.yml
@@ -163,11 +163,12 @@ jobs:
           echo "Installing CLI from: $CLI_URL"
           
           install_cli "$CLI_URL" "${{env.DOT_FORCE_DOWNLOAD}}"
+          
+          # Run the CLI push and capture exit code
           run_cli_push "${{github.workspace}}${{env.DOT_REPO_BASE_PATH}}" "${{ secrets.DOT_TOKEN }}" "${{ env.DOT_CLI_OPTS }}"
-          echo "exit-code=$exit_code" >> "$GITHUB_OUTPUT"
+          
+          # If we reach here, the command succeeded
+          echo "exit-code=0" >> "$GITHUB_OUTPUT"
           print_log
-          if [ $exit_code -ne 0 ]; then
-            echo "Error running dotCMS CLI"
-            exit 1
-          fi
+          echo "dotCMS CLI push completed successfully"
         shell: bash

--- a/tools/dotcms-cli/action/.github/workflows/main.yml
+++ b/tools/dotcms-cli/action/.github/workflows/main.yml
@@ -13,13 +13,13 @@ jobs:
         uses: actions/checkout@v4
 
       - name: Install xmllint
-        run: sudo apt-get update && sudo apt-get install -y libxml2-utils        
+        run: sudo apt-get update && sudo apt-get install -y libxml2-utils
 
       - name: Get dotCLI runner
         id: dotcli-runner
         run: |
           echo "::group::Get dotCLI runner"
-       
+          
           curl -o maven-metadata.xml "https://repo.dotcms.com/artifactory/libs-release-local/com/dotcms/dotcms-cli/maven-metadata.xml"
           LATEST_VERSION=$(xmllint --xpath "string(//versioning/latest)" maven-metadata.xml)
           URL="https://repo.dotcms.com/artifactory/libs-release-local/com/dotcms/dotcms-cli/${LATEST_VERSION}/dotcms-cli-${LATEST_VERSION}-runner.jar"
@@ -42,16 +42,37 @@ jobs:
       DOT_CREATE_WORKSPACE: ${{ vars.DOT_CREATE_WORKSPACE || 'true' }}
       # This is the CLI version to use, but we can always override this value
       DOT_CLI_JAR_DOWNLOAD_URL: ${{ vars.DOT_CLI_JAR_DOWNLOAD_URL || needs.dotcli-runner.outputs.url }}
+      # Support for version number only
+      DOT_CLI_VERSION: ${{ vars.DOT_CLI_VERSION }}
       # In case we want to force the download of the CLI jar
       DOT_FORCE_DOWNLOAD: ${{ vars.DOT_FORCE_DOWNLOAD || 'false' }}
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
 
       - name: Load .env file
         uses: aarcangeli/load-dotenv@v1.0.0
         with:
           if-file-not-found: 'warn'
+
+      - name: Resolve CLI URL
+        id: resolve-cli-url
+        run: |
+          # Check if DOT_CLI_VERSION is set (version number only)
+          if [ -n "${{ env.DOT_CLI_VERSION }}" ]; then
+            VERSION="${{ env.DOT_CLI_VERSION }}"
+            RESOLVED_URL="https://repo.dotcms.com/artifactory/libs-release-local/com/dotcms/dotcms-cli/${VERSION}/dotcms-cli-${VERSION}-runner.jar"
+            echo "::notice::Using specified version: $VERSION"
+            echo "::notice::Resolved URL: $RESOLVED_URL"
+            echo "resolved-url=$RESOLVED_URL" >> "$GITHUB_OUTPUT"
+          else
+            # Use the original URL (either from DOT_CLI_JAR_DOWNLOAD_URL or latest)
+            RESOLVED_URL="${{ env.DOT_CLI_JAR_DOWNLOAD_URL }}"
+            echo "::notice::Using configured or latest URL: $RESOLVED_URL"
+            echo "resolved-url=$RESOLVED_URL" >> "$GITHUB_OUTPUT"
+          fi
 
       - name: Print env vars
         run: |
@@ -59,7 +80,28 @@ jobs:
 
       - name: Get changes
         run: |
-          git diff --name-only ${{ github.event.before }} ${{ github.event.after }} > changed_files.txt
+          echo "::group::Detecting changed files"
+    
+          BEFORE="${{ github.event.before }}"
+          AFTER="${{ github.event.after }}"
+          
+          echo "Before: $BEFORE"
+          echo "After: $AFTER"
+          
+          if [ "$BEFORE" = "0000000000000000000000000000000000000000" ]; then
+          echo "::notice::First push - all files considered changed"
+          git ls-tree --name-only -r HEAD > changed_files.txt
+          elif git cat-file -e "$BEFORE" 2>/dev/null && git cat-file -e "$AFTER" 2>/dev/null; then
+          echo "::notice::Normal diff between valid commits"
+          git diff --name-only "$BEFORE" "$AFTER" > changed_files.txt
+          else
+          echo "::warning::One or both commits not accessible, falling back to current commit files"
+          git ls-tree --name-only -r HEAD > changed_files.txt
+          fi
+          
+          FILE_COUNT=$(wc -l < changed_files.txt)
+          echo "::notice::Files detected: $FILE_COUNT"
+          echo "::endgroup::"
 
       - name: List changed files
         run: |
@@ -111,9 +153,16 @@ jobs:
       - name: Run dotCMS CLI
         id: dot-push
         run: |
+          set -e  # Fail on any error
+          
           chmod +x ./.github/workflows/scripts/run-push.sh
           source ./.github/workflows/scripts/run-push.sh
-          install_cli "${{env.DOT_CLI_JAR_DOWNLOAD_URL}}" "${{env.DOT_FORCE_DOWNLOAD}}"
+          
+          # Use resolved URL from previous step
+          CLI_URL="${{ steps.resolve-cli-url.outputs.resolved-url }}"
+          echo "Installing CLI from: $CLI_URL"
+          
+          install_cli "$CLI_URL" "${{env.DOT_FORCE_DOWNLOAD}}"
           run_cli_push "${{github.workspace}}${{env.DOT_REPO_BASE_PATH}}" "${{ secrets.DOT_TOKEN }}" "${{ env.DOT_CLI_OPTS }}"
           echo "exit-code=$exit_code" >> "$GITHUB_OUTPUT"
           print_log

--- a/tools/dotcms-cli/action/.github/workflows/scripts/run-push.sh
+++ b/tools/dotcms-cli/action/.github/workflows/scripts/run-push.sh
@@ -13,7 +13,7 @@ SERVICES_FILE_CONTENT="
 
 _make_home(){
   if [ ! -d "$DOT_CLI_HOME" ]; then
-    mkdir $DOT_CLI_HOME
+    mkdir $DOT_CLI_HOME || return $?
   fi
 }
 
@@ -28,8 +28,8 @@ _get_CLI(){
     return
   fi
 
-  curl "$cli_release_download_url" -L -o ${DOT_CLI_HOME}${DOT_CLI_JAR}
-  chmod 777 "${DOT_CLI_HOME}${DOT_CLI_JAR}"
+  curl "$cli_release_download_url" -L -o ${DOT_CLI_HOME}${DOT_CLI_JAR} || return $?
+  chmod 777 "${DOT_CLI_HOME}${DOT_CLI_JAR}" || return $?
 
   #Check the size of the file
   file="${DOT_CLI_HOME}${DOT_CLI_JAR}" && \
@@ -37,7 +37,7 @@ _get_CLI(){
 
   if [ "$actual_size" -lt 1000000 ]; then
     echo "The file is too small to be the CLI, please check the version and try again"
-    exit 1
+    return 1
   fi
 }
 
@@ -48,8 +48,8 @@ _get_run_java_script(){
       echo "run-java.sh already exists, skipping download"
       return
    fi
-    curl https://repo1.maven.org/maven2/io/fabric8/run-java-sh/"${RUN_JAVA_VERSION}"/run-java-sh-"${RUN_JAVA_VERSION}"-sh.sh -o "${DOT_CLI_HOME}"run-java.sh
-    chmod 777 ${DOT_CLI_HOME}run-java.sh
+    curl https://repo1.maven.org/maven2/io/fabric8/run-java-sh/"${RUN_JAVA_VERSION}"/run-java-sh-"${RUN_JAVA_VERSION}"-sh.sh -o "${DOT_CLI_HOME}"run-java.sh || return $?
+    chmod 777 ${DOT_CLI_HOME}run-java.sh || return $?
 }
 
 _setup_CLI(){
@@ -60,14 +60,14 @@ _setup_CLI(){
    # All we need is a file with an active profile that matches the server we want to connect to in this case we are using default
    # If the directory does not exist we create it
     if [ ! -d "$DOT_SERVICES_HOME" ]; then
-      mkdir "$DOT_SERVICES_HOME"
+      mkdir "$DOT_SERVICES_HOME" || return $?
     else
        # If the directory exists we remove it as we could be updating the server url
-       rm -rf "$DOT_SERVICES_HOME"
-       mkdir "$DOT_SERVICES_HOME"
+       rm -rf "$DOT_SERVICES_HOME" || return $?
+       mkdir "$DOT_SERVICES_HOME" || return $?
     fi
     # Now generate the file
-    echo "$SERVICES_FILE_CONTENT" >> "$SERVICE_FILE";
+    echo "$SERVICES_FILE_CONTENT" >> "$SERVICE_FILE" || return $?
 
     export QUARKUS_LOG_FILE_PATH=$DOT_CLI_HOME"dotcms-cli.log"
 }
@@ -92,18 +92,17 @@ _run_cli_push(){
       export QUARKUS_LOG_FILE_PATH="$DOT_CLI_HOME"dotcms-cli.log
       cmd="bash /tmp/dot-cli/run-java.sh push $workspace_path $push_opts --token=$token"
       eval "$cmd"
-      export exit_code=$?
-      echo $exit_code
+      return $?
 }
 
 install_cli(){
     cli_release_download_url=$1
     force_download=$2
 
-    _make_home
-    _get_CLI "$cli_release_download_url" "$force_download"
-    _get_run_java_script "$force_download"
-    _setup_CLI
+    _make_home || return $?
+    _get_CLI "$cli_release_download_url" "$force_download" || return $?
+    _get_run_java_script "$force_download" || return $?
+    _setup_CLI || return $?
 }
 
 run_cli_push(){
@@ -114,6 +113,6 @@ run_cli_push(){
     echo "PUSH OPTS:"
     echo "$push_opts"
 
-    return_code=$(_run_cli_push "$workspace_path" "$token" "$push_opts")
-    echo "$return_code"
+    _run_cli_push "$workspace_path" "$token" "$push_opts"
+    return $?
 }

